### ISSUE TITLE
Improve pydocstyle (PEP257) compliance

### ISSUE THREAD

**Issue by peterjc** [biopython/biopython#1170]
Biopython aspires to follow Python's docstring style guidelines in PEP257 https://www.python.org/dev/peps/pep-0257/ (and use reStructuredText for the docstrings).

We run the automatic tool ``pydocstyle`` on the code as part of the automated TravisCI tests (also used on pull requests). Right now this is configured to ignore many style violations. For example, the new release of pydocstyle v2.0.0 added some extra checks, most of which I just disabled in order that our TravisCI builds pass - see https://github.com/biopython/biopython/commit/cc2ad72f1f237b3ff779b151fdf529abba1f3459

This issue is about (gradually) cleaning up PEP257 compliance issues as flagged by the latest version of the ``pydocstyle`` tool, and removing entries from the ``ignore`` setting in our ``.pydocstyle`` configuration file.

Note that in some cases we as a community might agree to continue to ignore some of these items.

To install/upgrade this tool, use something like this (or just ``pip`` if you are still on Python 2):

```
$ pip3 install --upgrade pydocstyle
```

The following command simples a useful way to see a tally table for picking off the rarest cases to fix them first. The cut is to remove the variable part of the some messages containing text from the offending line.

```
$ pydocstyle Bio/ BioSQL/ Tests/ Scripts/ Doc/ --ignore "" | grep "^        D" | cut -d "(" -f 1 | sort | uniq -c | sort -n
  11         D207: Docstring is under-indented
  12         D104: Missing docstring in public package
  18         D301: Use r""" if any backslashes in a docstring
  24         D214: Section is over-indented 
  33         D405: Section name should be properly capitalized 
  35         D211: No blank lines allowed before class docstring 
  40         D100: Missing docstring in public module
  43         D404: First word of the docstring should not be `This`
  63         D412: No blank lines allowed between a section header and its content 
  85         D103: Missing docstring in public function
 102         D212: Multi-line docstring summary should start at the first line
 104         D403: First word of the first line should be properly capitalized 
 108         D209: Multi-line docstring closing quotes should be on a separate line
 109         D210: No whitespaces allowed surrounding docstring text
 114         D413: Missing blank line after last section 
 118         D101: Missing docstring in public class
 123         D401: First line should be in imperative mood; try rephrasing 
 135         D208: Docstring is over-indented
 140         D406: Section name should end with a newline 
 142         D407: Missing dashed underline after section 
 187         D205: 1 blank line required between summary line and description 
 254         D200: One-line docstring should fit on one line with quotes 
 275         D105: Missing docstring in magic method
 283         D402: First line should not be the function's "signature"
 354         D204: 1 blank line required after class docstring 
 569         D400: First line should end with a period 
 590         D203: 1 blank line required before class docstring 
 623         D401: First line should be in imperative mood 
 746         D102: Missing docstring in public method
2344         D213: Multi-line docstring summary should start at the second line
```

Once you've picked an entry to work on, run ``pydocstyle`` again selecting just that entry, e.g.:

```
$ pydocstyle Bio/ BioSQL/ Tests/ Scripts/ Doc/ --select D104
...
```

I've marked this issue as "Easy Fix" in the hope of encouraging new-comers to try fixing some of these issues. If you would like to work on one of these (e.g. fixing the D104 entries), please comment below to avoid duplicated effort, before starting work and submitting a pull request. 

Note we will ask you to agree to dual licence your contributions under both the _Biopython License Agreement_ AND the _BSD 3-Clause License_ (see #898) and if you are you happy to be thanked by name in the ``NEWS.rst`` and ``CONTRIB.rst`` files (and under what name if this is not clear from your GitHub profile).

*Update*: You may find some of the modules have loads of issues, and instead it may make sense to work on all the docstring style or reStructuredText problems in that module at once. That's fine too - but again please comment on this issue to avoid duplication of effort.

Thanks!

*Update*: See #1203 for all the D1## entries about missing docstrings.

**Comment by morrme** [comment]
I would like to help with this, but as I'm new to your codebase, I'd need a little mentoring to get started.

**Comment by peterjc** [comment]
@morrme how about you try D211 then - it just means removing the blank lines which ``pydocstyle`` is complaining about?

You'll need to fork the Biopython repository, make a new branch, install and run ``pydocstyle``, fix the blank lines, commit the changes, push to your fork on GitHub, and then make a pull request. Does all of that make sense?

**Comment by morrme** [comment]
@peterjc yes, it does! i will get started with it and let you know if i run into any issues.

**Comment by morrme** [comment]
@peterjc all done! what is the next rule on your wishlist?

**Comment by peterjc** [comment]
@morrme Speedy! Thanks! Since you're not familiar with the codebase, I suggest you avoid the issues about missing docstrings. There are more blank line checks you could tackle if you like?

Or how about D301? It theory it should be easy if you're familiar with r"string" notation for raw strings to avoid the need to slash escape special characters, but I've not looked at why these are present here.

See also https://github.com/PyCQA/pydocstyle/issues/249 - some of these fixes can surely be automated which will save lots of people effort in the long run.

**Comment by morrme** [comment]
@peterjc sure, which ones are the blank line checks? D201-D205?
 
I can take a look at what is failing D301 as well. 

A lot of these can definitely be fixed automatically. I hope that issue you made gets some traction!

**Comment by souravsingh** [comment]
@peterjc I think the coala PyDocStyleBear could be useful for automated checks.

**Comment by peterjc** [comment]
@souravsingh That looks like a Coala wrapper for ``pydocstyle``, so wouldn't add anything directly, but using coala generally in our TravisCI/Tox style checks and/or git commit hook might be good. I suggest you comment on #883

**Comment by peterjc** [comment]
Cross reference https://github.com/PyCQA/pydocstyle/issues/250 which would make it easier to merge changes to ``.pydocstyle`` and comment why we still ignore some entries.

**Comment by peterjc** [comment]
Updated counts,

```
$ pydocstyle Bio/ BioSQL/ Tests/ Scripts/ Doc/ --ignore "" | grep "^        D" | cut -d "(" -f 1 | sort | uniq -c | sort -n
   3         D412: No blank lines allowed between a section header and its content 
  11         D207: Docstring is under-indented
  12         D104: Missing docstring in public package
  18         D301: Use r""" if any backslashes in a docstring
  23         D214: Section is over-indented 
  42         D100: Missing docstring in public module
  43         D404: First word of the docstring should not be `This`
  46         D403: First word of the first line should be properly capitalized 
  87         D103: Missing docstring in public function
  89         D212: Multi-line docstring summary should start at the first line
 106         D210: No whitespaces allowed surrounding docstring text
 118         D101: Missing docstring in public class
 123         D401: First line should be in imperative mood; try rephrasing 
 129         D208: Docstring is over-indented
 140         D406: Section name should end with a newline 
 142         D407: Missing dashed underline after section 
 155         D200: One-line docstring should fit on one line with quotes 
 174         D205: 1 blank line required between summary line and description 
 275         D105: Missing docstring in magic method
 282         D402: First line should not be the function's "signature"
 335         D204: 1 blank line required after class docstring 
 518         D400: First line should end with a period 
 621         D401: First line should be in imperative mood 
 626         D203: 1 blank line required before class docstring 
 748         D102: Missing docstring in public method
2260         D213: Multi-line docstring summary should start at the second line
```

D412 was annoying - it seemed to be clashing with valid RST content - so I reenabled it in https://github.com/biopython/biopython/commit/12ee1844c55fb4fb597ca817326ae74c38efbbd6

**Comment by morrme** [comment]
@peterjc I haven't gotten to them yet, but I am interested in the difference between D212 and D213.

**Comment by morrme** [comment]
But I am now working on D203 (`1 blank line required before class docstring`  )  and once I make changes, it violates D211 (No blank lines allowed before class docstring  ). So this is another pair of conflicting rules.

**Comment by peterjc** [comment]
I'd wondered from the description D2013 and D211 were mutually exclusive. Best ask on their issue tracker, perhaps they offer both and expect any one project to just neither, or pick one?

**Comment by peterjc** [comment]
See https://github.com/PyCQA/pydocstyle/issues/141 for D203 vs D211, by default D203 is ignored.

I think we should should aim to pass D211 (do not try to follow D203).

**Comment by morrme** [comment]
OK @peterjc I will undo my changes.  

What about D212 and D213? I think it is another case of having to pick between them because you cannot adhere to both.

**Comment by peterjc** [comment]
Yes, https://github.com/PyCQA/pydocstyle/issues/242 confirms D212 and D213 are mutually exclusive. It might be worth logging an issue for ``pydocstyle`` to error/warn when told to run mutually exclusive checks?

I think we should aim to pass D212 (do not try to follow D213).

To my eye this matches the https://www.python.org/dev/peps/pep-0257/#multi-line-docstrings example, and it reflects the current usage pattern in Biopython with 89 violations of D212, vs 2260 violations of D213 at last count.

**Comment by peterjc** [comment]
I'm doing D404 at the moment - it requires some editorial judgement ;)

**Comment by peterjc** [comment]
Updated numbers, ignoring D203 and D213 as discussed above:

```
$ pydocstyle Bio/ BioSQL/ Tests/ Scripts/ Doc/ --ignore D203,D213 | grep "^        D" | cut -d "(" -f 1 | sort | uniq -c | sort -n
   6         D412: No blank lines allowed between a section header and its content 
  11         D207: Docstring is under-indented
  12         D104: Missing docstring in public package
  18         D301: Use r""" if any backslashes in a docstring
  22         D214: Section is over-indented 
  39         D403: First word of the first line should be properly capitalized 
  42         D100: Missing docstring in public module
  61         D212: Multi-line docstring summary should start at the first line
  87         D103: Missing docstring in public function
  98         D210: No whitespaces allowed surrounding docstring text
 107         D208: Docstring is over-indented
 113         D401: First line should be in imperative mood; try rephrasing 
 115         D101: Missing docstring in public class
 132         D205: 1 blank line required between summary line and description 
 141         D406: Section name should end with a newline 
 143         D407: Missing dashed underline after section 
 144         D200: One-line docstring should fit on one line with quotes 
 152         D402: First line should not be the function's "signature"
 273         D105: Missing docstring in magic method
 333         D204: 1 blank line required after class docstring 
 457         D400: First line should end with a period 
 622         D401: First line should be in imperative mood 
 747         D102: Missing docstring in public method
1249         D415: Fails reStructuredText validation 
```

The mysterious new entry D415 is using an experimental hack to check the docstrings are valid reStructuredText via ``docutils``, see https://github.com/PyCQA/pydocstyle/pull/254

**Comment by morrme** [comment]
My next targets: D412,  D403 and D212 
then I'll look at D205, D406, D200

**Comment by peterjc** [comment]
@morrme Great - most of those look simple from their description, so good plan:

D403: First word of the first line should be properly capitalized 
D212: Multi-line docstring summary should start at the first line
D205: 1 blank line required between summary line and description
D200: One-line docstring should fit on one line with quotes 

These are going to depend on how exactly ``pydocstyle`` defines a section, which has puzzled me in the past, and may again conflict with passing RST validation:

D412: No blank lines allowed between a section header and its content
D406: Section name should end with a newline

**Comment by MarkusPiotrowski** [comment]
My feeling is that many of the errors overdo the job, e.g. 

>D404 	First word of the docstring should not be _This_
D405   Section name should be properly capitalized

You can't find anything about this in PEP257. What's the problem with _This_ as first word in a class docstring? Why should I capitalize any word in a section header? Don't want to speak about wrongly identified section headers...(just found out that section headers are identified by their text, e.g. 'Short summary', 'Parameters', 'See Also', etc)

These rules put much work on us and the results are questionably (see "See Also:" in ```Restriction.py```)

Thus I would suggest to focus on PEP257 by using ```-convention=pep257```. This is what ```flake8-docstrings``` does by default (= ommitting D203, D212, D213, D214, D215, D404, D405, D406, D407, D408, D409, D410,  D411)

**Comment by peterjc** [comment]
That explains the mystery of the section names, thanks @MarkusPiotrowski 

As to "This" as the first word, it does seem to be a good indicator that the docstring could be better worded.

Where do we stand if running ``pydocstyle -convention=pep257``, as that is a good standard to aim for.

**Comment by MarkusPiotrowski** [comment]
When running ```pydocstyle``` in the ```Bio``` folder:
```
pydocstyle --convention=pep257 --count
```
we got __3183__ issues.

**Comment by peterjc** [comment]
I've logged #1203 for the missing public docstrings (D1## errors).

**Comment by morrme** [comment]
@peterjc should i push the changed docstirngs or are we moving to make the other changes now?

**Comment by peterjc** [comment]
@morrme if you've got pending changes you've already made, you might as well put them into a pull request - even if we decide not to use them and/or it requires a manual merge.

Assuming #1202 is merged soon it will at a minimum mean differences in how the checks are done.

**Comment by peterjc** [comment]
I've just merged #1202 which means our TravisCI/Tox checks no longer use ``pydocstyle`` directly, but via ``flake8`` using the ``flake8-docstrings`` plugin.

This means ``.pydocstyle`` has gone away and instead the docstring settings are in ``Bio/.flake8``, ``BioSQL/.flake8``, etc. Having separate configurations for each folder is a short term solution allowing some rules to be more strictly enforced in some parts of the code than others.

This also means we can (for example) work on fixing a given docstring style issue on each of the top-level folders in turn, rather than all at once.

**Comment by morrme** [comment]
@peterjc I was wondering what would happen to the work we did here. 

So for now, what additional work needs to be done?

**Comment by peterjc** [comment]
Updated numbers ignoring D203 and D213 as discussed, and all the D1## missing docstring issues (see #1203),

```
$ pydocstyle Bio/ BioSQL/ Tests/ Scripts/ Doc/ --ignore D203,D213,D1 | grep "^        D" | cut -d "(" -f 1 | sort | uniq -c | sort -n
   6         D412: No blank lines allowed between a section header and its content 
  11         D207: Docstring is under-indented
  18         D301: Use r""" if any backslashes in a docstring
  22         D214: Section is over-indented 
  39         D403: First word of the first line should be properly capitalized 
  61         D212: Multi-line docstring summary should start at the first line
  95         D210: No whitespaces allowed surrounding docstring text
 103         D208: Docstring is over-indented
 108         D401: First line should be in imperative mood; try rephrasing 
 128         D205: 1 blank line required between summary line and description 
 141         D406: Section name should end with a newline 
 143         D402: First line should not be the function's "signature"
 143         D407: Missing dashed underline after section 
 144         D200: One-line docstring should fit on one line with quotes 
 330         D204: 1 blank line required after class docstring 
 445         D400: First line should end with a period 
 606         D401: First line should be in imperative mood 
```

BioSQL, Scripts and Doc seem to be OK.

It looks like a small job to resolve these issues in the ``Tests/`` folder:

```
$ pydocstyle Tests/ --ignore D203,D213,D1 | grep "^        D" | cut -d "(" -f 1 | sort | uniq -c | sort -n
   1         D200: One-line docstring should fit on one line with quotes 
   2         D204: 1 blank line required after class docstring 
   3         D401: First line should be in imperative mood 
   7         D400: First line should end with a period 
  16         D403: First word of the first line should be properly capitalized
```

While for Bio there are still plenty of things to look at:

```
$ pydocstyle Bio/ --ignore D203,D213,D1 | grep "^        D" | cut -d "(" -f 1 | sort | uniq -c | sort -n
   6         D412: No blank lines allowed between a section header and its content 
  11         D207: Docstring is under-indented
  18         D301: Use r""" if any backslashes in a docstring
  22         D214: Section is over-indented 
  23         D403: First word of the first line should be properly capitalized 
  61         D212: Multi-line docstring summary should start at the first line
  95         D210: No whitespaces allowed surrounding docstring text
 103         D208: Docstring is over-indented
 108         D401: First line should be in imperative mood; try rephrasing 
 128         D205: 1 blank line required between summary line and description 
 141         D406: Section name should end with a newline 
 143         D200: One-line docstring should fit on one line with quotes 
 143         D402: First line should not be the function's "signature"
 143         D407: Missing dashed underline after section 
 328         D204: 1 blank line required after class docstring 
 438         D400: First line should end with a period 
 603         D401: First line should be in imperative mood
```

**Comment by morrme** [comment]
@peterjc How do you feel about `D401: First line should be in imperative mood ` , given some of the other discussion here?

**Comment by peterjc** [comment]
Personally I think ``D401: First line should be in imperative mood`` is worth fixing but as a low priority - and I recognise not everyone would agree so it may not end up on our enforced list. While sometimes it is a trivial fix (e.g. ``Returns ...`` vs ``Return ...``), in other cases it requires a bit more thought - and having fresh eyes on any docstring should be a good thing.

All the 'section header' ones are also potentially questionable, e.g. D412 and D407.

These are probably all clear cut (and some ought to be easy to automate):

D200: One-line docstring should fit on one line with quotes 
D204: 1 blank line required after class docstring 
D210: No whitespaces allowed surrounding docstring text
D212: Multi-line docstring summary should start at the first line
D402: First line should not be the function's "signature"
D400: First line should end with a period

**Comment by peterjc** [comment]
I'm currently doing ``D207: Docstring is under-indented``

**Comment by morrme** [comment]
OK then I'll go for `D200: One-line docstring should fit on one line with quotes`

**Comment by peterjc** [comment]
New numbers for ``Bio/``

```
$ pydocstyle Bio/ --ignore D1,D203,D213,D407,D412 | grep "^        D" | cut -d "(" -f 1 | sort | uniq -c | sort -n
  16         D301: Use r""" if any backslashes in a docstring
  21         D403: First word of the first line should be properly capitalized 
  22         D214: Section is over-indented 
  86         D210: No whitespaces allowed surrounding docstring text
  98         D208: Docstring is over-indented
 102         D205: 1 blank line required between summary line and description 
 105         D401: First line should be in imperative mood; try rephrasing 
 131         D402: First line should not be the function's "signature"
 142         D406: Section name should end with a newline 
 387         D400: First line should end with a period 
 593         D401: First line should be in imperative mood 
```


I'm not 100% sure why but this wrongly tells me the Tests folder is now clear:

```
$ pydocstyle Tests/ --ignore D1,D203,D213,D407,D412
(clear)
```

vs:

```
$ pydocstyle Tests/*.py --select D200,D202,D204,D403 | grep "^        D" | cut -d "(" -f 1 | sort | uniq -c sort -n
  33         D403: First word of the first line should be properly capitalized 
  67         D204: 1 blank line required after class docstring 
 202         D202: No blank lines allowed after function docstring 
 217         D200: One-line docstring should fit on one line with quotes
```

Thus when I tightened the TravisCI settings in https://github.com/biopython/biopython/commit/39d4f6297524bf7a1fa5d8b8914d88e26cd8144a I had to undo it in https://github.com/biopython/biopython/commit/44fe007c802d618c48a5d51b578fa58f91dbbe78

**Comment by peterjc** [comment]
Ah, one mystery solved - by default ``pydocstyle`` ignores ``test_*.py``,

```
$ pydocstyle --help
...
  --match=<pattern>     check only files that exactly match <pattern> regular
                        expression; default is --match='(?!test_).*\.py' which
                        matches files that don't start with 'test_' but end
                        with '.py'
...
```

This seems to do the trick:

```
$ pydocstyle Tests/ --ignore D1,D203,D213,D407,D412 --match '.+\.py'
...
```

Giving:

```
$ pydocstyle Tests/ --ignore D1,D203,D213,D407,D412 --match '.+\.py' | grep "^        D" | cut -d "(" -f 1 | sort | uniq -c | sort -n
   6         D212: Multi-line docstring summary should start at the first line
  15         D401: First line should be in imperative mood; try rephrasing 
  31         D401: First line should be in imperative mood 
  33         D403: First word of the first line should be properly capitalized 
  67         D204: 1 blank line required after class docstring 
 137         D300: Use """triple double quotes""" 
 202         D202: No blank lines allowed after function docstring 
 217         D200: One-line docstring should fit on one line with quotes 
 939         D400: First line should end with a period 
```

In any case, the ``Bio`` entries are more pressing - the main thing the docstrings in the ``Tests/`` folder are used for is the test names in verbose output when running the tests.

**Comment by morrme** [comment]
In the above output for `D401`(for both Tests and Bio), there are 2 different entries for it in the same result set. 

`205` is tricky; some of the docstrings dont have natural "summary lines"; your corrections from the last commit were reasonable, but there are others that are moreso sample inputs, etc. 

This also leads to issues with `D400: First line should end with a period ` ,  `D402: First line should not be the function's "signature"` , `D403: First word of the first line should be properly capitalized `

**Comment by peterjc** [comment]
@morrme - Its probably just a quirk of how my piped command line truncating the variable part of the message which means D401 gets two separate entries.

I think we do want to comply with D205 (docstring should start with a one line summary), although doing this isn't something we can automate - some of these will be hard without knowing the module.

Of the remaining items these look most straightforward (and potentially could be automated):
* D200: One-line docstring should fit on one line with quotes 
* D202: No blank lines allowed after function docstring 
* D204: 1 blank line required after class docstring

**Comment by peterjc** [comment]
I've done ``D214: Section is over-indented`` with 57771d301df503410fabc819b3500a4ac237594d (which took a long time as I tried to tackle a lot of docstring style changes in that module) and aca5162ae6b44900e4d4d3fd31b5fdba57c7774c

Strangely D214 is not in ``Bio/.flake8`` yet seems to be ignored by default...

**Comment by morrme** [comment]
OK where are we now?

**Comment by peterjc** [comment]
```
$ pydocstyle Bio/ --ignore D1,D203,D213 | grep "^        D" | cut -d "(" -f 1 | sort | uniq -c | sort -n
  12         D208: Docstring is over-indented
  16         D301: Use r""" if any backslashes in a docstring
  17         D412: No blank lines allowed between a section header and its content 
  21         D403: First word of the first line should be properly capitalized 
  76         D402: First line should not be the function's "signature"
  95         D205: 1 blank line required between summary line and description 
 104         D401: First line should be in imperative mood; try rephrasing 
 136         D406: Section name should end with a newline 
 138         D407: Missing dashed underline after section 
 313         D400: First line should end with a period 
 566         D401: First line should be in imperative mood 
1054         D415: Fails reStructuredText validation
```

This is using experimental pydocstyle hack to check the docstrings are valid reStructuredText via docutils, D415, see PyCQA/pydocstyle#254

We're ignoring D203 in favour of passing D211.

We're ignoring D213 in favour of passing D212.

Of these I would focus on D208 (indentation) and these first-line points:

* D200: One-line docstring should fit on one line with quotes
* D202: No blank lines allowed after function docstring
* D204: 1 blank line required after class docstring

For D412, how does this look for ``Bio/GA/Organism.py``:

```
Represent a single individual in a population.                                                                                              
                                                                                                                                                   
Attributes:                                                                                                                                    
 - genome -- The genome of the organism. This is a Bio.MutableSeq                                                                               
   object that has the sequence of the genome, and the alphabet                                                                                 
   describing all elements that can be a part of the genome.                                                                                    
 - fitness -- The calculate fitness of the organism. This fitness is                                                                            
   based on the last time it was calculated using the fitness_calculator.                                                                       
   So... the fitness could potentially be out of date with the real genome                                                                      
   if you are not careful to recalculate it after changes with                                                                                  
   recalculate_fitness()                                                                                                                        

```

vs the current version which breaks D412:

```
Represent a single individual in a population.                                                                                              

Attributes:

- genome -- The genome of the organism. This is a Bio.MutableSeq                                                                               
  object that has the sequence of the genome, and the alphabet
  describing all elements that can be a part of the genome.
- fitness -- The calculate fitness of the organism. This fitness is
  based on the last time it was calculated using the fitness_calculator.
  So... the fitness could potentially be out of date with the real genome
  if you are not careful to recalculate it after changes with
  recalculate_fitness()

```

Using http://rst.ninjs.org/ they render subtly differently, but either looks fine to me.

(This is not that important as the next goal would be to pick and use a standard style for the argument and attribute lists, perhaps following the NumPy docstring example https://github.com/numpy/numpy/blob/master/doc/HOWTO_DOCUMENT.rst.txt#documenting-classes - or the RST field lists as used in Bio.Phylo which epydoc understands and checks)

**Comment by morrme** [comment]
Your `Organism.py` replacement looks good. 
200/202/204 do not appear in your Bio/ output. Should I look at these in Tests/ ?

I will look at the Phylo style again but the Numpy style is clear.

**Comment by peterjc** [comment]
If you want to work on the ``Tests/`` sure - but here the docstrings are less important as they are not user facing.

Sorry, I'd copy-pasted already solved line-one issues, how about these next - all of which will require looking at the context and selecting a good one line summary:

* D402: First line should not be the function's "signature"
* D205: 1 blank line required between summary line and description
* D400: First line should end with a period

Note we don't want to blindly "fix" D205 and D400 where there currently isn't a one line summary.

**Comment by peterjc** [comment]
Current numbers,

```
$ pydocstyle Bio/ --ignore D1,D203,D213 | grep "^        D" | cut -d "(" -f 1 | sort | uniq -c | sort -n
  15         D301: Use r""" if any backslashes in a docstring
  23         D403: First word of the first line should be properly capitalized 
  57         D402: First line should not be the function's "signature"
  85         D205: 1 blank line required between summary line and description 
  94         D406: Section name should end with a newline 
  96         D407: Missing dashed underline after section 
  99         D401: First line should be in imperative mood; try rephrasing 
 276         D400: First line should end with a period 
 514         D401: First line should be in imperative mood 
1104         D415: Fails reStructuredText validation 
```

As before, ``D415: Fails reStructuredText validation`` is from an experimental fork of ``pydocstyle``, see https://github.com/PyCQA/pydocstyle/pull/254

**Comment by peterjc** [comment]
Current numbers as of commit 1abacd6739468226b5f20d4557cfa9b5401ebbff which follows me doing a bunch of work on #1221 and fixing some related docstring issues along the way:

```
$ pydocstyle Bio/ --ignore D1,D203,D213 | grep "^        D" | cut -d "(" -f 1 | sort | uniq -c | sort -n
  14         D301: Use r""" if any backslashes in a docstring
  14         D403: First word of the first line should be properly capitalized 
  39         D402: First line should not be the function's "signature"
  59         D205: 1 blank line required between summary line and description 
  65         D401: First line should be in imperative mood; try rephrasing 
 105         D406: Section name should end with a newline 
 107         D407: Missing dashed underline after section 
 179         D400: First line should end with a period 
 452         D401: First line should be in imperative mood 
```

**Comment by morrme** [comment]
You chose to un-ignore D407. Is that intentional?

**Comment by peterjc** [comment]
Do you mean not ignoring D407 in the ``pydocstyle`` output on this issue, or in the code? I've just been repeating the same command the last few times, not sure why back in May I was more inclusive.

Markus made a good argument that we shouldn't worry about D407:

>Thus I would prefer to focus on PEP257 by using -convention=pep257. This is what flake8-docstrings does by default (= ommitting D203, D212, D213, D214, D215, D404, D405, D406, D407, D408, D409, D410, D411).

Right now I would still focus on:

* D205: 1 blank line required between summary line and description
* D400: First line should end with a period
* D402: First line should not be the function's "signature"

Also still worth looking at but, as you know, trickier that it looks at first glance due to doctests,

* D301: Use r""" if any backslashes in a docstring

**Comment by morrme** [comment]
I agree with ignoring D407, I just noticed it in the output of your most recent issue count.

**Comment by peterjc** [comment]
New numbers, ignoring the controversial D406 and D407:

```
$ pydocstyle Bio/ --ignore D1,D203,D213,D406,D407 | grep "^        D" | cut -d "(" -f 1 | sort | uniq -c | sort -n
  14         D301: Use r""" if any backslashes in a docstring
  39         D402: First line should not be the function's "signature"
  48         D205: 1 blank line required between summary line and description 
  62         D401: First line should be in imperative mood; try rephrasing 
 156         D400: First line should end with a period 
 413         D401: First line should be in imperative mood 
```

**Comment by peterjc** [comment]
I've addressed ``D402`` now, there were a few false positives which was curious.

New numbers, ignoring the controversial D406 and D407:

```
$ pydocstyle Bio/ --ignore D1,D203,D213,D406,D407 | grep "^        D" | cut -d "(" -f 1 | sort | uniq -c | sort -n
  11         D301: Use r""" if any backslashes in a docstring
  46         D205: 1 blank line required between summary line and description 
  62         D401: First line should be in imperative mood; try rephrasing 
 124         D400: First line should end with a period 
 410         D401: First line should be in imperative mood 
```

**Comment by peterjc** [comment]
Little change in the latest numbers, still ignoring the controversial D406 and D407:

```
$ pydocstyle Bio/ --ignore D1,D203,D213,D406,D407 | grep "^        D" | cut -d "(" -f 1 | sort | uniq -c | sort -n
pydocstyle --version
  12         D301: Use r""" if any backslashes in a docstring
  46         D205: 1 blank line required between summary line and description 
  62         D401: First line should be in imperative mood; try rephrasing 
 122         D400: First line should end with a period 
 395         D401: First line should be in imperative mood 
$ pydocstyle --version
2.1.1
```

**Comment by peterjc** [comment]
I'm working on D205 right now...

**Comment by peterjc** [comment]
I've started work on D400 now.

**Comment by peterjc** [comment]
Pull request #1462 would fix and enforce D205 and D400 via our TravisCI/Tox style check, which would leave missing docstrings (#1203) and the following:

```
$ pydocstyle Bio/ --ignore D1,D203,D213,D406,D407 | grep "^        D" | cut -d "(" -f 1 | sort | uniq -c | sort -n
  12         D301: Use r""" if any backslashes in a docstring
  62         D401: First line should be in imperative mood; try rephrasing 
 375         D401: First line should be in imperative mood
```

To be honest, while I would like to follow D401 imperative mood, it seems a very low priority.

I have started work on D301, cross reference discussion on #1173.

**Comment by souravsingh** [comment]
I can work on the D401 if it is needed. If not, we can ignore the error.

**Comment by peterjc** [comment]
Once #1173 is merged I think we can close this issue :)

@saketkc If you want to try doing a few D401 cases, sure - give it a try. Please start on those modules you have contributed to and are most familiar with. Thank you.

**Comment by peterjc** [comment]
Current numbers,

```
$ pydocstyle Bio/ --ignore D1,D203,D213,D406,D407 | grep "^        D" | cut -d "(" -f 1 | sort | uniq -c | sort -n
  62         D401: First line should be in imperative mood; try rephrasing 
 354         D401: First line should be in imperative mood 
```

i.e. 416 cases of D401 which I have logged as #1473 as a low priority issue.

Closing this as most of the pydocstyle checks now pass. Horray - thank you everyone who contributed to this effort!
